# Spark API Exercises

In [1]:
# imports
import pyspark
import pandas as pd
import numpy as np

from pyspark.sql.functions import *

In [2]:
# sparks fly
spark = pyspark.sql.SparkSession.builder.getOrCreate()

# ------------------------------------------------------------------------
## Exercise 1:

Within your `codeup-data-science` directory, create a new repo named `spark-exercises`. This will be where you do your work for this module. Create a repository on GitHub with the same name, and link your local repository to GitHub.

Save this work in your `spark-exercises` repo. Then add, commit, and push your changes.

Create a jupyter notebook or python script named `spark101` for this exercise.

Create a spark data frame that contains your favorite programming languages.

- Create a dataframe with one column named `language`
> Hint: Start with a pandas dataframe. Maybe use a dictionary?
- View the schema of the dataframe
- Output the shape of the dataframe
- Show the first 5 records in the dataframe

# ------------------------------------------------------------------------

Create a dataframe with one column named `language`

In [3]:
# create pandas dataframe
languages = pd.DataFrame(dict(language = ['python', 'java', 'html', 'c', 'r', 'php', 'scala'],
                            studied = ['yes', 'no', 'no', 'no', 'no', 'no', 'no'])
                        )
# a look at the dataframe
languages

language studied
0   python     yes
1     java      no
2     html      no
3        c      no
4        r      no
5      php      no
6    scala      no

In [4]:
# create spark dataframe from pandas dataframe
df = spark.createDataFrame(languages)

# view dataframe columns and datatypes
df

DataFrame[language: string, studied: string]

View the schema of the dataframe

In [5]:
# view dataframe schema
df.printSchema()

root
 |-- language: string (nullable = true)
 |-- studied: string (nullable = true)



Ouput the shape of the dataframe

In [6]:
print('DataFrame shape: ', df.count(), ' x ', len(df.columns))

DataFrame shape:  7  x  2


Show the first five records in the dataframe

In [7]:
# show first five records of dataframe
df.show(5)

+--------+-------+
|language|studied|
+--------+-------+
|  python|    yes|
|    java|     no|
|    html|     no|
|       c|     no|
|       r|     no|
+--------+-------+
only showing top 5 rows



# ------------------------------------------------------------------------
## Exercise 2:

Load the `mpg` dataset as a spark dataframe.

a. Create 1 column of output that contains a message like the one below for each record:

    The 1999 audi a4 has a 4 cylinder engine.

> Hint: You will need to concatenate values that already exist in the data with string literals

b. Transform the trans column so that it only contains either manual or auto.

> Hint: Consider spark string methods and `when().otherwise()` chaining
# ------------------------------------------------------------------------

In [8]:
from pydataset import data
mpg = spark.createDataFrame(data('mpg'))
mpg.show()

+------------+------------------+-----+----+---+----------+---+---+---+---+-------+
|manufacturer|             model|displ|year|cyl|     trans|drv|cty|hwy| fl|  class|
+------------+------------------+-----+----+---+----------+---+---+---+---+-------+
|        audi|                a4|  1.8|1999|  4|  auto(l5)|  f| 18| 29|  p|compact|
|        audi|                a4|  1.8|1999|  4|manual(m5)|  f| 21| 29|  p|compact|
|        audi|                a4|  2.0|2008|  4|manual(m6)|  f| 20| 31|  p|compact|
|        audi|                a4|  2.0|2008|  4|  auto(av)|  f| 21| 30|  p|compact|
|        audi|                a4|  2.8|1999|  6|  auto(l5)|  f| 16| 26|  p|compact|
|        audi|                a4|  2.8|1999|  6|manual(m5)|  f| 18| 26|  p|compact|
|        audi|                a4|  3.1|2008|  6|  auto(av)|  f| 18| 27|  p|compact|
|        audi|        a4 quattro|  1.8|1999|  4|manual(m5)|  4| 18| 26|  p|compact|
|        audi|        a4 quattro|  1.8|1999|  4|  auto(l5)|  4| 16| 25|  p|c

In [9]:
# view columns and data types
mpg.printSchema()

root
 |-- manufacturer: string (nullable = true)
 |-- model: string (nullable = true)
 |-- displ: double (nullable = true)
 |-- year: long (nullable = true)
 |-- cyl: long (nullable = true)
 |-- trans: string (nullable = true)
 |-- drv: string (nullable = true)
 |-- cty: long (nullable = true)
 |-- hwy: long (nullable = true)
 |-- fl: string (nullable = true)
 |-- class: string (nullable = true)



a. Create 1 column of output that contains a message like the one below for each record:

    The 1999 audi a4 has a 4 cylinder engine.

In [10]:
# write out pyspark concatenation of literal (lit) strings and column (col) values
mpg.select(concat(lit('The '),
                  col('year'),
                  lit(' '),
                  col('manufacturer'),
                  lit(' '),
                  col('model'),
                  lit(' has a '), 
                  col('cyl'),
                  lit(' cylinder engine.')
                 ).alias('summary_description')
          ).\
show(11, truncate = False)

+-------------------------------------------------+
|summary_description                              |
+-------------------------------------------------+
|The 1999 audi a4 has a 4 cylinder engine.        |
|The 1999 audi a4 has a 4 cylinder engine.        |
|The 2008 audi a4 has a 4 cylinder engine.        |
|The 2008 audi a4 has a 4 cylinder engine.        |
|The 1999 audi a4 has a 6 cylinder engine.        |
|The 1999 audi a4 has a 6 cylinder engine.        |
|The 2008 audi a4 has a 6 cylinder engine.        |
|The 1999 audi a4 quattro has a 4 cylinder engine.|
|The 1999 audi a4 quattro has a 4 cylinder engine.|
|The 2008 audi a4 quattro has a 4 cylinder engine.|
|The 2008 audi a4 quattro has a 4 cylinder engine.|
+-------------------------------------------------+
only showing top 11 rows



b. Transform the trans column so that it only contains either manual or auto.

> Hint: Consider spark string methods and `when().otherwise()` chaining

In [11]:
# review dataframe
mpg.show(2)

+------------+-----+-----+----+---+----------+---+---+---+---+-------+
|manufacturer|model|displ|year|cyl|     trans|drv|cty|hwy| fl|  class|
+------------+-----+-----+----+---+----------+---+---+---+---+-------+
|        audi|   a4|  1.8|1999|  4|  auto(l5)|  f| 18| 29|  p|compact|
|        audi|   a4|  1.8|1999|  4|manual(m5)|  f| 21| 29|  p|compact|
+------------+-----+-----+----+---+----------+---+---+---+---+-------+
only showing top 2 rows



In [12]:
# register mpg table with spark
mpg.createOrReplaceTempView('mpg')

In [13]:
# create mysql query to grab columns to fit needs
mpg = spark.sql(
'''
SELECT manufacturer, model, displ, cyl, LEFT(trans, 4) as trans, drv, cty, hwy, fl, class
FROM mpg
''')
mpg.show()

+------------+------------------+-----+---+-----+---+---+---+---+-------+
|manufacturer|             model|displ|cyl|trans|drv|cty|hwy| fl|  class|
+------------+------------------+-----+---+-----+---+---+---+---+-------+
|        audi|                a4|  1.8|  4| auto|  f| 18| 29|  p|compact|
|        audi|                a4|  1.8|  4| manu|  f| 21| 29|  p|compact|
|        audi|                a4|  2.0|  4| manu|  f| 20| 31|  p|compact|
|        audi|                a4|  2.0|  4| auto|  f| 21| 30|  p|compact|
|        audi|                a4|  2.8|  6| auto|  f| 16| 26|  p|compact|
|        audi|                a4|  2.8|  6| manu|  f| 18| 26|  p|compact|
|        audi|                a4|  3.1|  6| auto|  f| 18| 27|  p|compact|
|        audi|        a4 quattro|  1.8|  4| manu|  4| 18| 26|  p|compact|
|        audi|        a4 quattro|  1.8|  4| auto|  4| 16| 25|  p|compact|
|        audi|        a4 quattro|  2.0|  4| manu|  4| 20| 28|  p|compact|
|        audi|        a4 quattro|  2.0

In [14]:
# test functionality
mpg.select(col('*')).show()

+------------+------------------+-----+---+-----+---+---+---+---+-------+
|manufacturer|             model|displ|cyl|trans|drv|cty|hwy| fl|  class|
+------------+------------------+-----+---+-----+---+---+---+---+-------+
|        audi|                a4|  1.8|  4| auto|  f| 18| 29|  p|compact|
|        audi|                a4|  1.8|  4| manu|  f| 21| 29|  p|compact|
|        audi|                a4|  2.0|  4| manu|  f| 20| 31|  p|compact|
|        audi|                a4|  2.0|  4| auto|  f| 21| 30|  p|compact|
|        audi|                a4|  2.8|  6| auto|  f| 16| 26|  p|compact|
|        audi|                a4|  2.8|  6| manu|  f| 18| 26|  p|compact|
|        audi|                a4|  3.1|  6| auto|  f| 18| 27|  p|compact|
|        audi|        a4 quattro|  1.8|  4| manu|  4| 18| 26|  p|compact|
|        audi|        a4 quattro|  1.8|  4| auto|  4| 16| 25|  p|compact|
|        audi|        a4 quattro|  2.0|  4| manu|  4| 20| 28|  p|compact|
|        audi|        a4 quattro|  2.0

In [15]:
# select trans column, and an aliased column of transmission, showing only 'automatic' or 'manual'
mpg.select(when(mpg.trans == 'auto', 'auto')
           .otherwise('manual')
           .alias('trans')
          )\
.show()

+------+
| trans|
+------+
|  auto|
|manual|
|manual|
|  auto|
|  auto|
|manual|
|  auto|
|manual|
|  auto|
|manual|
|  auto|
|  auto|
|manual|
|  auto|
|manual|
|  auto|
|  auto|
|  auto|
|  auto|
|  auto|
+------+
only showing top 20 rows



In [16]:
# pull original complete table
mpg = spark.createDataFrame(data('mpg'))

In [17]:
# ALTERNATE SOLUTION
mpg.select(mpg.trans,
           when(length(mpg.trans) < 10, 'automatic')
           .otherwise('manual')
           .alias('transmission')
          )\
.show()

+----------+------------+
|     trans|transmission|
+----------+------------+
|  auto(l5)|   automatic|
|manual(m5)|      manual|
|manual(m6)|      manual|
|  auto(av)|   automatic|
|  auto(l5)|   automatic|
|manual(m5)|      manual|
|  auto(av)|   automatic|
|manual(m5)|      manual|
|  auto(l5)|   automatic|
|manual(m6)|      manual|
|  auto(s6)|   automatic|
|  auto(l5)|   automatic|
|manual(m5)|      manual|
|  auto(s6)|   automatic|
|manual(m6)|      manual|
|  auto(l5)|   automatic|
|  auto(s6)|   automatic|
|  auto(s6)|   automatic|
|  auto(l4)|   automatic|
|  auto(l4)|   automatic|
+----------+------------+
only showing top 20 rows



In [18]:
# ALTERNATE SOLUTION ('transform' the trans column)
mpg.select(mpg.manufacturer,
           mpg.model,
           mpg.displ,
           mpg.cyl,
           when(length(mpg.trans) < 10, 'auto')
           .otherwise('manual')
           .alias('transmission'),
           mpg.drv,
           mpg.cty,
           mpg.hwy,
           mpg.fl,
           col('class')
          )\
.show()

+------------+------------------+-----+---+------------+---+---+---+---+-------+
|manufacturer|             model|displ|cyl|transmission|drv|cty|hwy| fl|  class|
+------------+------------------+-----+---+------------+---+---+---+---+-------+
|        audi|                a4|  1.8|  4|        auto|  f| 18| 29|  p|compact|
|        audi|                a4|  1.8|  4|      manual|  f| 21| 29|  p|compact|
|        audi|                a4|  2.0|  4|      manual|  f| 20| 31|  p|compact|
|        audi|                a4|  2.0|  4|        auto|  f| 21| 30|  p|compact|
|        audi|                a4|  2.8|  6|        auto|  f| 16| 26|  p|compact|
|        audi|                a4|  2.8|  6|      manual|  f| 18| 26|  p|compact|
|        audi|                a4|  3.1|  6|        auto|  f| 18| 27|  p|compact|
|        audi|        a4 quattro|  1.8|  4|      manual|  4| 18| 26|  p|compact|
|        audi|        a4 quattro|  1.8|  4|        auto|  4| 16| 25|  p|compact|
|        audi|        a4 qua

# ------------------------------------------------------------------------
## Exercise 3: 

Load the `tips` dataset as a spark dataframe.

a. What percentage of observations are smokers?
> Hint: `.groupBy()` and `.withColumn()` are useful functions here

b. Create a column that contains the tip percentage
> Hint: `.withColumn()` is useful here

c. Calculate the average tip percentage for each combination of sex and smoker.
> Hint: Chain additional functions off the answer to part b 

# ------------------------------------------------------------------------

In [19]:
# create spark dataframe from tips data set and show top 20 rows
tips = spark.createDataFrame(data('tips'))
tips.show()

+----------+----+------+------+---+------+----+
|total_bill| tip|   sex|smoker|day|  time|size|
+----------+----+------+------+---+------+----+
|     16.99|1.01|Female|    No|Sun|Dinner|   2|
|     10.34|1.66|  Male|    No|Sun|Dinner|   3|
|     21.01| 3.5|  Male|    No|Sun|Dinner|   3|
|     23.68|3.31|  Male|    No|Sun|Dinner|   2|
|     24.59|3.61|Female|    No|Sun|Dinner|   4|
|     25.29|4.71|  Male|    No|Sun|Dinner|   4|
|      8.77| 2.0|  Male|    No|Sun|Dinner|   2|
|     26.88|3.12|  Male|    No|Sun|Dinner|   4|
|     15.04|1.96|  Male|    No|Sun|Dinner|   2|
|     14.78|3.23|  Male|    No|Sun|Dinner|   2|
|     10.27|1.71|  Male|    No|Sun|Dinner|   2|
|     35.26| 5.0|Female|    No|Sun|Dinner|   4|
|     15.42|1.57|  Male|    No|Sun|Dinner|   2|
|     18.43| 3.0|  Male|    No|Sun|Dinner|   4|
|     14.83|3.02|Female|    No|Sun|Dinner|   2|
|     21.58|3.92|  Male|    No|Sun|Dinner|   2|
|     10.33|1.67|Female|    No|Sun|Dinner|   3|
|     16.29|3.71|  Male|    No|Sun|Dinne

In [20]:
# view dataframe schema
tips.printSchema()

root
 |-- total_bill: double (nullable = true)
 |-- tip: double (nullable = true)
 |-- sex: string (nullable = true)
 |-- smoker: string (nullable = true)
 |-- day: string (nullable = true)
 |-- time: string (nullable = true)
 |-- size: long (nullable = true)



In [21]:
print('DataFrame shape: ', tips.count(), ' x ', len(tips.columns))

DataFrame shape:  244  x  7


a. What percentage of observations are smokers?
> Hint: `.groupBy()` and `.withColumn()` are useful functions here

In [22]:
# get smoker fruit rollups
tips.rollup('smoker').count().show()

+------+-----+
|smoker|count|
+------+-----+
|  null|  244|
|    No|  151|
|   Yes|   93|
+------+-----+



In [23]:
# get smoker fruit rollups with percentage
tips.rollup('smoker').count().withColumn('percent',
                                         round(col('count') / tips.count() * 100)
                                        )\
.show()

+------+-----+-------+
|smoker|count|percent|
+------+-----+-------+
|  null|  244|  100.0|
|    No|  151|   62.0|
|   Yes|   93|   38.0|
+------+-----+-------+



38% of observations are smokers.

b. Create a column that contains the tip percentage
> Hint: `.withColumn()` is useful here

In [24]:
# produce og df supplemented w/ 'tip_percent' column, the concatenation of the rounded percentage and the '%' sign
tips.withColumn('tip_percent', concat(round(tips.tip / tips.total_bill * 100, 2), lit('%'))).show()

+----------+----+------+------+---+------+----+-----------+
|total_bill| tip|   sex|smoker|day|  time|size|tip_percent|
+----------+----+------+------+---+------+----+-----------+
|     16.99|1.01|Female|    No|Sun|Dinner|   2|      5.94%|
|     10.34|1.66|  Male|    No|Sun|Dinner|   3|     16.05%|
|     21.01| 3.5|  Male|    No|Sun|Dinner|   3|     16.66%|
|     23.68|3.31|  Male|    No|Sun|Dinner|   2|     13.98%|
|     24.59|3.61|Female|    No|Sun|Dinner|   4|     14.68%|
|     25.29|4.71|  Male|    No|Sun|Dinner|   4|     18.62%|
|      8.77| 2.0|  Male|    No|Sun|Dinner|   2|     22.81%|
|     26.88|3.12|  Male|    No|Sun|Dinner|   4|     11.61%|
|     15.04|1.96|  Male|    No|Sun|Dinner|   2|     13.03%|
|     14.78|3.23|  Male|    No|Sun|Dinner|   2|     21.85%|
|     10.27|1.71|  Male|    No|Sun|Dinner|   2|     16.65%|
|     35.26| 5.0|Female|    No|Sun|Dinner|   4|     14.18%|
|     15.42|1.57|  Male|    No|Sun|Dinner|   2|     10.18%|
|     18.43| 3.0|  Male|    No|Sun|Dinne

c. Calculate the average tip percentage for each combination of sex and smoker.
> Hint: Chain additional functions off the answer to part b 

In [25]:
# produce tip percent column, then group by sex and pivot by smoker status, and aggregate by average tip percent values
tips.withColumn('tip_percent', (tips.tip / tips.total_bill)
               ).groupby('sex').pivot('smoker').agg(round(mean('tip_percent') * 100, 2))\
.show()

+------+-----+-----+
|   sex|   No|  Yes|
+------+-----+-----+
|Female|15.69|18.22|
|  Male|16.07|15.28|
+------+-----+-----+



# ------------------------------------------------------------------------
## Exercise 4:

Use the seattle weather dataset referenced in the lesson to answer the questions below.

- Convert the temperatures to fahrenheit.
- Which month has the most rain, on average?
- Which year was the windiest?
- What is the most frequent type of weather in January?
- What is the average high and low temperature on sunny days in July in 2013 and 2014?
- What percentage of days were rainy in q3 of 2015?
- For each year, find what percentage of days it rained (had non-zero precipitation).

In [26]:
# import data
from vega_datasets import data

# get data into pandas dataframe and assign new column for string-type dates
weather = data.seattle_weather().assign(date = lambda df: df.date.astype(str))

# get pandas dataframe into spark dataframe
weather = spark.createDataFrame(weather)

# view dataframe
weather.show()

+----------+-------------+--------+--------+----+-------+
|      date|precipitation|temp_max|temp_min|wind|weather|
+----------+-------------+--------+--------+----+-------+
|2012-01-01|          0.0|    12.8|     5.0| 4.7|drizzle|
|2012-01-02|         10.9|    10.6|     2.8| 4.5|   rain|
|2012-01-03|          0.8|    11.7|     7.2| 2.3|   rain|
|2012-01-04|         20.3|    12.2|     5.6| 4.7|   rain|
|2012-01-05|          1.3|     8.9|     2.8| 6.1|   rain|
|2012-01-06|          2.5|     4.4|     2.2| 2.2|   rain|
|2012-01-07|          0.0|     7.2|     2.8| 2.3|   rain|
|2012-01-08|          0.0|    10.0|     2.8| 2.0|    sun|
|2012-01-09|          4.3|     9.4|     5.0| 3.4|   rain|
|2012-01-10|          1.0|     6.1|     0.6| 3.4|   rain|
|2012-01-11|          0.0|     6.1|    -1.1| 5.1|    sun|
|2012-01-12|          0.0|     6.1|    -1.7| 1.9|    sun|
|2012-01-13|          0.0|     5.0|    -2.8| 1.3|    sun|
|2012-01-14|          4.1|     4.4|     0.6| 5.3|   snow|
|2012-01-15|  

In [27]:
weather.printSchema()

root
 |-- date: string (nullable = true)
 |-- precipitation: double (nullable = true)
 |-- temp_max: double (nullable = true)
 |-- temp_min: double (nullable = true)
 |-- wind: double (nullable = true)
 |-- weather: string (nullable = true)



In [28]:
print('DataFrame shape: ', weather.count(), ' x ', len(weather.columns))

DataFrame shape:  1461  x  6


Convert the temperatures to Fahrenheit.

In [29]:
# assign dataframe to dataframe supplemented w/ columns that represent max & min temps converted to fahrenheit
weather = weather\
.withColumn('max_fahrenheit', round(weather.temp_max * 9 / 5 + 32, 2))\
.withColumn('min_fahrenheit', round(weather.temp_min * 9 / 5 + 32, 2))

weather.show()    # show dataframe

+----------+-------------+--------+--------+----+-------+--------------+--------------+
|      date|precipitation|temp_max|temp_min|wind|weather|max_fahrenheit|min_fahrenheit|
+----------+-------------+--------+--------+----+-------+--------------+--------------+
|2012-01-01|          0.0|    12.8|     5.0| 4.7|drizzle|         55.04|          41.0|
|2012-01-02|         10.9|    10.6|     2.8| 4.5|   rain|         51.08|         37.04|
|2012-01-03|          0.8|    11.7|     7.2| 2.3|   rain|         53.06|         44.96|
|2012-01-04|         20.3|    12.2|     5.6| 4.7|   rain|         53.96|         42.08|
|2012-01-05|          1.3|     8.9|     2.8| 6.1|   rain|         48.02|         37.04|
|2012-01-06|          2.5|     4.4|     2.2| 2.2|   rain|         39.92|         35.96|
|2012-01-07|          0.0|     7.2|     2.8| 2.3|   rain|         44.96|         37.04|
|2012-01-08|          0.0|    10.0|     2.8| 2.0|    sun|          50.0|         37.04|
|2012-01-09|          4.3|     9

Which month has the most rain, on average?

In [30]:
# Get the year and month columns, group by month and then year for all combos, aggregate the 
# total precipitation for each combo and sort in descending order
weather.withColumn('month', month('date'))\
.withColumn('year', year('date'))\
.groupBy('month', 'year')\
.agg(sum('precipitation').alias('total_rains'))\
.sort(col('total_rains').desc())\
.show()

+-----+----+------------------+
|month|year|       total_rains|
+-----+----+------------------+
|   12|2015| 284.5000000000001|
|    3|2014|240.00000000000003|
|   11|2015|             212.6|
|   11|2012|             210.5|
|    3|2012|             183.0|
|   12|2012|             174.0|
|    1|2012|173.29999999999998|
|   10|2014|             171.5|
|   10|2012|170.29999999999998|
|    9|2013|156.79999999999998|
|    2|2014|155.20000000000002|
|    4|2013|149.60000000000002|
|    2|2015|134.19999999999996|
|   11|2014|             123.1|
|   10|2015|122.39999999999998|
|   12|2014|121.79999999999998|
|    3|2015|113.49999999999997|
|    4|2014|106.10000000000001|
|    1|2013|105.69999999999997|
|   11|2013|              96.3|
+-----+----+------------------+
only showing top 20 rows



In [31]:
# Get year and month columns, group by month and year for all combos, aggregate all precipitation for 
# each combo, then group by month and aggregate on average precipitation, sort in descending order and grab the 
# first record
weather.withColumn('month', month('date'))\
.withColumn('year', year('date'))\
.groupBy('month', 'year')\
.agg(sum('precipitation').alias('total_rains'))\
.groupby('month')\
.agg(mean('total_rains').alias('average_rains'))\
.sort(col('average_rains').desc())\
.first()

Row(month=11, average_rains=160.625)

Which year was the windiest?

In [32]:
# Get year column, group by year, aggregate the sum of all wind values, sort in descending order and view the 
# first 3 records
weather.withColumn('year', year('date'))\
.groupby('year')\
.agg(sum('wind').alias('total_wind'))\
.sort(col('total_wind').desc())\
.head(3)

[Row(year=2012, total_wind=1244.7),
 Row(year=2014, total_wind=1236.5000000000007),
 Row(year=2015, total_wind=1153.3000000000002)]

2012 was the windiest year.

What is the most frequent type of weather in January?

In [33]:
# Get a month column going, filter for january, group by weather and aggregate on counts of different weather types
weather.withColumn('month' , month('date'))\
.filter(col('month') == 1)\
.groupby('weather')\
.agg(count('weather').alias('weather_counts'))\
.sort(col('weather_counts').desc())\
.show()

+-------+--------------+
|weather|weather_counts|
+-------+--------------+
|    fog|            38|
|   rain|            35|
|    sun|            33|
|drizzle|            10|
|   snow|             8|
+-------+--------------+



Fog is the most frequent weather type in January.

What is the average high and low temperature on sunny days in July in 2013 and 2014?

In [34]:
# add day, month and year columns to reduce redundancy
weather = weather.withColumn('day', dayofmonth('date')).withColumn('month', month('date')).withColumn('year', year('date'))
weather.show(5)

+----------+-------------+--------+--------+----+-------+--------------+--------------+---+-----+----+
|      date|precipitation|temp_max|temp_min|wind|weather|max_fahrenheit|min_fahrenheit|day|month|year|
+----------+-------------+--------+--------+----+-------+--------------+--------------+---+-----+----+
|2012-01-01|          0.0|    12.8|     5.0| 4.7|drizzle|         55.04|          41.0|  1|    1|2012|
|2012-01-02|         10.9|    10.6|     2.8| 4.5|   rain|         51.08|         37.04|  2|    1|2012|
|2012-01-03|          0.8|    11.7|     7.2| 2.3|   rain|         53.06|         44.96|  3|    1|2012|
|2012-01-04|         20.3|    12.2|     5.6| 4.7|   rain|         53.96|         42.08|  4|    1|2012|
|2012-01-05|          1.3|     8.9|     2.8| 6.1|   rain|         48.02|         37.04|  5|    1|2012|
+----------+-------------+--------+--------+----+-------+--------------+--------------+---+-----+----+
only showing top 5 rows



In [35]:
# get supplemented df schema
weather.printSchema()

root
 |-- date: string (nullable = true)
 |-- precipitation: double (nullable = true)
 |-- temp_max: double (nullable = true)
 |-- temp_min: double (nullable = true)
 |-- wind: double (nullable = true)
 |-- weather: string (nullable = true)
 |-- max_fahrenheit: double (nullable = true)
 |-- min_fahrenheit: double (nullable = true)
 |-- day: integer (nullable = true)
 |-- month: integer (nullable = true)
 |-- year: integer (nullable = true)



In [36]:
# Filter results for only sunny days in July 2013 & 2014, aggregate average max and min temperatures
weather.filter(weather.month == 7)\
.filter((weather.year < 2015) & (weather.year > 2012))\
.filter(weather.weather == 'sun')\
.agg(mean('max_fahrenheit').alias('highest_avg'),
     mean('min_fahrenheit').alias('lowest_avg')
    ).show()

+-----------------+-----------------+
|      highest_avg|       lowest_avg|
+-----------------+-----------------+
|80.29192307692308|57.52884615384615|
+-----------------+-----------------+



For July's sunny days in 2013 and 2014, the highest average temperature was 80.3 degrees Fahrenheit, and the lowest average temperature was 57.5 degrees Fahrenheit.

What percentage of days were rainy in q3 of 2015?

In [37]:
# Filter results for July, August and September 2015 (3rd quarter), turn rain column into 1s and 0s, aggregate
# by mean to get proportion of rainy days
weather.filter(weather.year == 2015)\
.filter((weather.month == 7) | (weather.month == 8) | (weather.month == 9))\
.select(when(weather.weather == 'rain', 1).otherwise(0).alias('rain'))\
.agg(mean('rain'))\
.show()

+--------------------+
|           avg(rain)|
+--------------------+
|0.021739130434782608|
+--------------------+



2.17% of days were rainy in quarter 3 of 2015

For each year, find what percentage of days it rained (had non-zero precipitation).

In [38]:
# Select precipitation column, change any values greater than 0 in precipitation to 1 and the rest 0, select the
# year column, group by the year, and then get proportions by aggregating on the mean of the new precipitation
# column values
weather.select(when(weather.precipitation > 0, 1).otherwise(0).alias('rainy'), weather.year)\
.groupby(weather.year)\
.agg(round(mean('rainy') * 100, 2).alias('percent_rainy'))\
.show()

+----+-------------+
|year|percent_rainy|
+----+-------------+
|2012|        48.36|
|2013|        41.64|
|2014|         41.1|
|2015|        39.45|
+----+-------------+

